py3status/modules/usbguard.py

```diff
@@ -0,0 +1,212 @@
+                    old_key = new_key.replace('_', '-')
+                    if old_key in event[4][4]:
+                        if event[4][4][old_key]:
+                            device[new_key] = event[4][4][old_key]
+                        else:
+                            device[new_key] = None
+                self.parent.data[usbguard_id] = device
+        else:
+            if self.parent.data[usbguard_id]:
+                del self.parent.data[usbguard_id]
+
+        self.parent.new_data = self.parent._manipulate_devices(
+            self.parent.data)
+        self.parent.py3.update()
+
+    # on policy change signal
+    def _on_devices_policy_changed(self, *event):
+        # TODO: send notification with action
+        usbguard_id = event[4][0]
+        device_perms = event[4][3]
+        actions = ['allow id', 'reject id']
+        for action in actions:
+            if action in device_perms:
+                if self.parent.data[usbguard_id]:
+                    del self.parent.data[usbguard_id]
+                    break
+
+        self.parent.new_data = self.parent._manipulate_devices(
+            self.parent.data)
+        self.parent.py3.update()
+
+    def run(self):
+        while not self.parent.killed.is_set():
+            self.parent._init_dbus()
+            self.parent.dbus.subscribe(
+                object=self.parent.dbus_devices,
+                signal='DevicePresenceChanged',
+                signal_fired=self._on_devices_presence_changed)
+
+            self.parent.dbus.subscribe(
+                object=self.parent.dbus_devices,
+                signal='DevicePolicyChanged',
+                signal_fired=self._on_devices_policy_changed)
+
+            self.loop = GLib.MainLoop()
+            self.loop.run()
+
+
+class Py3status:
+    """
+    """
+    button_allow = 1
+    button_block = 3
+    button_reject = None
+    format = u'[USBGuard: {name}]'
+    format_separator = u' | '
+
+    def _init_dbus(self):
+        self.dbus_interface = 'org.usbguard'
+        self.dbus_devices = '/org/usbguard/Devices'
+        self.dbus = SystemBus()
+        self.error = None
+        try:
+            self.proxy = self.dbus.get(self.dbus_interface)
+        except:
+            self.error = Exception(STRING_USBGUARD_DBUS)
+
+    def _init_placeholders(self):
+        available_placeholders = [
+            'id', 'name', 'via_port', 'hash', 'parent_hash', 'serial',
+            'with_interface'
+        ]
+
+        placeholders = {}
+
+        for placeholder in available_placeholders:
+            if self.py3.format_contains(self.format, placeholder):
+                placeholders[placeholder] = None
+
+        placeholders['usbguard_id'] = None
+        return placeholders
+
+    def post_config_hook(self):
+        self._init_dbus()
+        self.data = {}
+        self.new_data = []
+
+        self.placeholders = self._init_placeholders()
+
+        self.permanant_rule = False
```

Comment:
Hardcoded! This could be enhanced more if you want it.
* Make it a per-device permanent rule.
* Make this boolean and toggleable.
* I think `format_permanent_rule` might be needed here so we can assign icons or different words instead of True/False... and is separated because of placeholder + index.

20 Slytherin points.